src/openforms/registrations/contrib/objects_api/tests/test_api_endpoints.py

```diff
@@ -0,0 +1,178 @@
+from unittest.mock import patch
+
+import requests_mock
+from rest_framework import status
+from rest_framework.reverse import reverse_lazy
+from rest_framework.test import APITestCase
+from zgw_consumers.models import Service
+
+from openforms.accounts.tests.factories import StaffUserFactory, UserFactory
+
+from ..models import ObjectsAPIConfig
+
+
+@requests_mock.Mocker()
```

Comment:
please use VCR instead

Reply:
See https://github.com/open-formulieren/open-forms/pull/3860#discussion_r1484368602, in my opinion VCR/requests_mock isn't the solution

Reply:
I don't see how that addresses the actual data going over the wire.

The behaviour that you are testing is:

1. when calling our own endpoint, actually the data from the objecttypes API is returned
2. this data itself is a stripped down set of properties from the upstream API
3. our endpoint does not allow extracting this data if the service is not properly configured

A mock client class does not do any of that.

I have near-zero interest in unit-tests for this functionality, I don't particularly care about the client implementation either. What I require is some integration-test functionality: from my frontend, I can retrieve the available object types and their versions from my configured objecttypes API. All the rest are implementation details.

Reply:
The objecttypes client itself is tested [here](https://github.com/open-formulieren/open-forms/pull/3847/files#diff-02ce813114e188adba916eec0086bcb7d784a9ae84462385bb346352ac71f8d1), hence my original comment on duplication.

This test was here to test the API endpoint, where the client used is an implementation detail, and I feel in this case we want to test that the serializier is doing its job, the permissions are correct, etc.

This test file is almost a duplicate of this [test file](https://github.com/open-formulieren/open-forms/blob/master/src/openforms/registrations/contrib/camunda/tests/test_api_endpoints.py), where the client data is mocked, so that's why I went the same way.

If this should be an integration test, then I should remove  [here](https://github.com/open-formulieren/open-forms/pull/3847/files#diff-02ce813114e188adba916eec0086bcb7d784a9ae84462385bb346352ac71f8d1)

Reply:
The Camunda integration is very old from before when we even knew VCR exists - that's no longer a good reference.

I would indeed rather see the existing client tests be "lifted up" to the functionality that we're actually interested in - i.e. our own API endpoints. The access control tests of course also need to happen, but those are easy to shortcut & you can also deliberately include a mis-configuration to see how our own API endpoints react (a missing/bad API key for example).

I don't think I agree on the duplication part. yes, some test code is duplicated to set up the service/client, but they serve to distinct goals:

* do our own API endpoints return what is needed?
* is the client implementation solid enough?

The client tests are a lower-level aspect to achieve the first item, but they have the advantage that they're easier (less-integration-y) to debug breakage with. If both the client + api endpoint tests, it should be easier to dive into the client tests to resolve the problem

I've seen too many issues with stale mocks, incorrect mocks or mocks that correctly implement the API spec but the service itself doesn't actually behave according to its documentation. That's why I strongly prefer using VCR for any external HTTP traffic instead of coming up with my own edge cases/things that can go wrong. E.g. with a mock client, you wouldn't think of returning an HTTP 500 for the `/api/v2/objecttypes/:bad-uuid/versions` endpoint, right?